src/Icicle/Source/ToCore/Base.hs

```diff
@@ -91,6 +94,77 @@ freshly f
         return (f n', n')
 
 
+-- | Apply a function to some arguments, depending on the arguments' "possibility".
+-- If any of the arguments are "possibly", unwrap their possibilities and
+-- rewrap the entire expression in a possibility.
+--
+-- The function type here is the "raw" return type of the function.
+-- That is, its modality/possibility is unaffected by those of
+-- its arguments.
+--
+-- This means for division, the modality should be "Possible",
+-- but for addition the modality should be "Definitely"
+-- even if the addition is applied to Possible arguments,
+-- and the actual application's return type would be Possible.
+--
+applyPossibles  :: C.Exp n
+                -> UniverseType
+                -> [(C.Exp n, UniverseType)]
+                -> ConvertM a n (C.Exp n)
+applyPossibles f returns xts
+ -- Go through xts, building up the list of actual expressions
+ -- to apply as arguments, and keep track of whether we've
+ -- seen any possibles so far.
+ = go xts [] False
+
+ where
+  -- At the end of xts, apply the function to the arguments
+  go [] args anyPossibles
+   -- If we've seen any possibles, and the raw function itself
+   -- returns a definitely, we need to wrap the result
+   -- in a "Some" constructor.
+   | anyPossibles
+   , Definitely <- universePossibility $ universe returns
+   = return $ mkSome $ X.makeApps f args
+   -- Otherwise, we have seen no possibles or the raw function
+   -- itself returns possibles, so we can leave the return as-is.
+   | otherwise
+   = return $          X.makeApps f args
+
+  -- For each entry of xts..
+  go ((x,u):xts') args anyPossibles
```

Comment:
am I a bad functional programmer because I wrote this with recursion instead of a "fold"?